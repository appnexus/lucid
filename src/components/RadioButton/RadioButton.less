@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

@RadioButton-size: 22px;
@RadioButton-animation-timing: 0.1s linear;

@RadioButton-container-offset: @size-XXS;
@RadioButton-container-radius: 8px;

@RadioButton-dot-offset: 8px;
@RadioButton-dot-radius: @size-XXS;

@RadioButton-glow-offset: @size-XXS;
@RadioButton-glow-radius: 8px;
@RadioButton-glow-scale: scale(1.375);

.lucid-RadioButton {
	cursor: pointer;
	display: inline-block;
	height: @RadioButton-size;
	position: relative;
	user-select: none;
	width: @RadioButton-size;

	&-is-disabled {
		cursor: not-allowed;
	}

	&-native {
		.opacity(0);
	}

	&-visualization-container {
		background-color: white;
		border-radius: @RadioButton-container-radius + 1;
		border: solid 1px rgba(0, 0, 0, 0.3);
		.box-sizing();
		height: @RadioButton-container-radius * 2;
		left: @RadioButton-container-offset;
		position: absolute;
		top: @RadioButton-container-offset;
		width: @RadioButton-container-radius * 2;
	}

	&-visualization-dot {
		background-color: white;
		border-radius: @RadioButton-dot-radius;
		.box-sizing();
		height: @RadioButton-dot-radius * 2;
		left: @RadioButton-dot-offset;
		.opacity(0);
		position: absolute;
		top: @RadioButton-dot-offset;
		transition: @RadioButton-animation-timing;
		width: @RadioButton-dot-radius * 2;
	}

	&-visualization-glow {
		background-color: @color-primary;
		border-radius: @RadioButton-glow-radius;
		.box-sizing();
		height: @RadioButton-glow-radius * 2;
		left: @RadioButton-glow-offset;
		.opacity(0);
		position: absolute;
		top: @RadioButton-glow-offset;
		transition: @RadioButton-animation-timing;
		width: @RadioButton-glow-radius * 2;
	}

	&:hover &-visualization-container {
		border-color: @color-primary;
		transition: @RadioButton-animation-timing;
	}

	&:hover &-visualization-glow {
		.opacity();
		transform: @RadioButton-glow-scale;
	}

	&-is-disabled &-visualization-container,
	&-is-disabled:hover &-visualization-container {
		background-color: rgba(0, 0, 0, 0.05);
		border-color: rgba(0, 0, 0, 0.1);
	}

	&-is-disabled:hover &-visualization-glow {
		.opacity(0);
		transform: none;
	}

	&-is-disabled&-is-selected &-visualization-container {
		.opacity();
	}

	&-is-disabled&-is-selected:hover &-visualization-container,
	&-is-selected &-visualization-container,
	&:active &-visualization-container {
		background-color: @color-primary;
		border-color: @color-primary;
	}

	&-is-selected &-visualization-dot {
		.opacity(1);
	}

	&-native:focus ~ &-visualization-glow {
		.opacity();
		transform: @RadioButton-glow-scale;
	}
}
