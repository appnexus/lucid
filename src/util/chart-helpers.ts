import _ from 'lodash';
import * as d3TimeFormat from 'd3-time-format';
import * as d3Time from 'd3-time';

export type Collection = Array<{ [key: string]: any }>;
type Fields = string[] | string;

/**
 * stackByFields
 *
 * D3's `stack` groups each series' data together but we sometimes we want the
 * stacked groups to remain grouped as in the original normalized data. This
 * function helps achieve that.
 *
 * @param {object[]} collection - normalized data you want to operate on
 * @param {string[]} fields - fields to pluck off for the y data
 * @return {array[]} - array of arrays, one for row in the original `collection`
 */
export function stackByFields(
	collection: Collection,
	fields: Fields
): Array<Array<[number, number]>> {
	const fieldsArray = _.castArray(fields);

	return _.map(collection, (d) => {
		return _.reduce<Fields, Array<[number, number]>>(
			fieldsArray,
			(acc, field) => {
				const dataPoint: number = _.get(d, field, 0);

				if (_.isEmpty(acc)) {
					return acc.concat([[0, dataPoint]]);
				}

				const maybeLast = _.last(_.last(acc));
				const last = maybeLast === undefined ? 0 : maybeLast;

				return acc.concat([[last, last + dataPoint]]);
			},
			[]
		);
	});
}

/**
 * extractFields
 *
 * This will return the data in a similar format to stackByFields but without
 * the stacking.
 *
 * @param {object[]} collection - normalized data you want to operate on
 * @param {string[]} fields - fields to pluck off for the y data
 * @return {array[]} - array of arrays, one for each field
 */
export function extractFields(
	collection: Collection,
	fields: Fields,
	minDomainValue: number = 0
): Array<Array<[number, number]>> {
	const fieldsArray = _.castArray(fields);

	return _.map(collection, (d) => {
		return _.map(fieldsArray, (field) => [minDomainValue, _.get(d, field, 0)]);
	});
}

/**
 * groupByFields
 *
 * This will return the data in a similar format to d3Shape.stack
 * but without the stacking of the data.
 *
 * @param {object[]} collection - normalized data you want to operate on
 * @param {string[]} fields - fields to pluck off for the y data
 * @return {array[]} - array of arrays, one for each field
 */
export function groupByFields(collection: Collection, fields: Fields) {
	const fieldsArray = _.castArray(fields);

	return _.map(fieldsArray, (field) => {
		return _.map(collection, field);
	});
}

/**
 * byFields
 *
 * Takes a collection of data and returns an array of all the fields off that
 * collection.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {array}
 */
export function byFields(collection: Collection, fields: Fields) {
	const fieldArray = _.castArray(fields);

	return _.reduce(
		fieldArray,
		(acc: object[], field) => {
			return acc.concat(_.map(collection, field));
		},
		[]
	);
}

/**
 * nearest
 *
 * Divide and conquer algorithm that helps find the nearest element to `value`
 * in `nums`
 *
 * @param {number[]} nums - sorted array of numbers to search through
 * @param {number} value - value you're trying to locate the nearest array element for
 * @return {number} - the nearest array element to the value
 */
export function nearest(nums: number[], value: number): number | undefined {
	if (nums.length < 2) {
		return _.first(nums);
	}

	if (nums.length === 2) {
		return value > (nums[0] + nums[1]) / 2 ? nums[1] : nums[0];
	}

	const mid = nums.length >>> 1;

	return nums[mid] > value
		? nearest(nums.slice(0, mid + 1), value)
		: nearest(nums.slice(mid), value);
}

/**
 * minByFields
 *
 * Returns the minimum element from a collection by a set of fields.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function minByFields(collection: Collection, fields: Fields) {
	return _.min(byFields(collection, fields));
}

/**
 * maxByFields
 *
 * Returns the maximum element from a collection by a set of fields.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function maxByFields(collection: Collection, fields: Fields) {
	return _.max(byFields(collection, fields));
}

/**
 * maxByFieldsStacked
 *
 * Returns the max sum of a set of fields from a collection
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function maxByFieldsStacked(collection: Collection, fields: Fields) {
	const fieldArray = _.castArray(fields);

	const sums = _.reduce(
		collection,
		(acc: number[], item) => {
			return acc.concat(_.sum(_.toArray(_.pick(item, fieldArray))));
		},
		[]
	);

	return _.max(sums);
}

/**
 * discreteTicks
 *
 * Returns `count` evenly spaced, representative values from the `array`.
 *
 * @param {array} array
 * @param {number} size - should be greater than 1
 * @return {array}
 */
export function discreteTicks<T>(
	array: T[],
	count: number | undefined | null = null
) {
	if (!array || _.isNil(count) || array.length <= count) {
		return array;
	}

	const step = (array.length - 1) / Math.max(1, count - 1);

	return _.reduce(
		_.times(count),
		(acc: T[], n) => {
			return acc.concat(array[Math.round(n * step)]);
		},
		[]
	);
}

/**
 * transformFromCenter
 *
 * Scaling paths from their center is tricky. This function
 * helps do that be generating a translate/scale transform
 * string with the correct numbers.
 *
 * @param {number} x - the x data point where you want the path to be centered at
 * @param {number} y - the y data point where you want the path to be centered at
 * @param {number} xCenter - the x coordinate of the center of the path you're trying to transform
 * @param {number} yCenter - the x coordinate of the center of the path you're trying to transform
 * @param {number} scale - number to scale to, 2 would be 2x bigger
 * @return {string} - transform string
 */
export function transformFromCenter(
	x: number,
	y: number,
	xCenter: number,
	yCenter: number,
	scale: number
) {
	return `translate(${(1 - scale) * xCenter + (x - xCenter)}, ${
		(1 - scale) * yCenter + (y - yCenter)
	}) scale(${scale})`;
}

const FORMAT_MILLISECOND = d3TimeFormat.timeFormat('.%L');
const FORMAT_SECOND = d3TimeFormat.timeFormat(':%S');
const FORMAT_MINUTE = d3TimeFormat.timeFormat('%I:%M');
const FORMAT_HOUR = d3TimeFormat.timeFormat('%I %p');
const FORMAT_DAY = d3TimeFormat.timeFormat('%a %d');
const FORMAT_WEEK = d3TimeFormat.timeFormat('%b %d');
const FORMAT_MONTH = d3TimeFormat.timeFormat('%b');
const FORMAT_YEAR = d3TimeFormat.timeFormat('%Y');

/**
 * formatDate
 *
 * This function was written to be used for tick formatting with d3 time
 * scales.
 *
 * @param {date} date - input date
 * @return {string} - formatted date
 */
export function formatDate(date: Date): string {
	return (
		d3Time.timeSecond(date) < date
			? FORMAT_MILLISECOND
			: d3Time.timeMinute(date) < date
			? FORMAT_SECOND
			: d3Time.timeHour(date) < date
			? FORMAT_MINUTE
			: d3Time.timeDay(date) < date
			? FORMAT_HOUR
			: d3Time.timeMonth(date) < date
			? d3Time.timeWeek(date) < date
				? FORMAT_DAY
				: FORMAT_WEEK
			: d3Time.timeYear(date) < date
			? FORMAT_MONTH
			: FORMAT_YEAR
	)(date);
}
