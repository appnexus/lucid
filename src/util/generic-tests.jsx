import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, {
	getDefaultProps = _.noop,
	selector = null,
	exemptFunctionProps = [],
} = {}) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const rootStyle = rootWrapper.prop('style');
			assert(_.every(style, (val, key) => val === rootStyle[key]), 'root style must contain passed styles');
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, className => {
				assert(_.includes(className, `lucid-${Component.displayName}`), `${className} must be scoped`);
			});
		});

		it('should only use onX convention for function proptypes', () => {

			_.forEach(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				const isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				const isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`

				if (isProbablyFunction) {
					assert(isAny || _.startsWith(key, 'on') || _.includes(exemptFunctionProps, key), `${key} must follow onX convention`);
				}

			});

		});

		// Only run this test if it's a public component
		if (!Component._lucidIsPrivate) {
			it('should be available as an exported module from index.js', () => {
				assert(lucid[Component.displayName]);
			});
		}
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM('[icon]', () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialProp: expectedSpecialProp,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { props: { specialProp }, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(specialProp, expectedSpecialProp, 'incorrect or missing specialProp');
		});
	});
}
