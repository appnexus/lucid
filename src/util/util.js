import _ from 'lodash';
import React from 'react';

export const isDevMode = (function isReactInDev() {
	try {
		React.createClass({});
	} catch(e) {
		if (e.message.indexOf('render') !== 0) {
			return true;  // A nice, specific error message
		}
	}

	return false;  // should never happen, but play it safe.
})();

export const isNode = typeof process === 'object' && process.title === 'node';

export const logger = (function() {
	return checkIsDev() ? {
		log,
		warn,
		error
	} : {
		log: _.noop,
		warn: _.noop,
		error: _.noop
	};
})();

export function checkIsDev() {
	return isDevMode &&
		typeof window !== 'undefined' &&
		!isNode &&
		typeof console !== 'undefined';
}

function log(...args) {

	console.log(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function warn(...args) {

	console.warn(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function error(...args) {

	console.error(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}
