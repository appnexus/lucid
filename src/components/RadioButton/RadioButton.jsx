import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import { lucidClassNames } from '../../util/style-helpers';
import {
	createClass as createReactClass,
	omitProps,
} from '../../util/component-types';

const cx = lucidClassNames.bind('&-RadioButton');
const { bool, func, object, string } = PropTypes;

/**
 * {"categories": ["controls", "toggles"]}
 *
 * This is a toggle -- a component that is in one of two particular states at
 * any given moment in time -- that features a custom visualization of the
 * native radio button control to reflect its current state.
 *
 * The radio button is different from a standard toggle in that when it is in
 * the selected state user events do not cause it to toggle to the unselected
 * state so the `onSelect` function is called only when `isSelected` is false.
 *
 * It uses a hidden native radio button control under the hood but leverages
 * other HTML elements to visualize its state.
 *
 * Any props that are not explicitly defined in `propTypes` are spread onto the
 * native control.
 */
const RadioButton = createReactClass({
	displayName: 'RadioButton',
	propTypes: {
		/**
		 * Appended to the component-specific class names set on the root
		 * element.
		 */
		className: string,

		/**
		 * Indicates whether the component should appear and act disabled by
		 * having a "greyed out" palette and ignoring user interactions.
		 */
		isDisabled: bool,

		/**
		 * Indicates that the component is in the "selected" state when true
		 * and in the "unselected" state when false.
		 */
		isSelected: bool,

		/**
		 * Called when the user clicks on the component or when they press the
		 * space key while the component is in focus, and only called when the
		 * component is in the unselected state.
		 *
		 * Signature: `(true, { event, props }) => {}`
		 */
		onSelect: func,

		/**
		 * Passed through to the root element.
		 */
		style: object,
	},

	getDefaultProps() {
		return {
			isDisabled: false,
			isSelected: false,
			onSelect: _.noop,
		};
	},

	componentDidMount() {
		this.nativeElement = this.refs.nativeElement;
	},

	handleSpanClick(e) {
		e.preventDefault();
	},

	render() {
		const {
			className,
			isDisabled,
			isSelected,
			style,
			...passThroughs
		} = this.props;

		return (
			<span
				className={cx(
					'&',
					{
						'&-is-disabled': isDisabled,
						'&-is-selected': isSelected,
					},
					className
				)}
				onClick={this.handleClicked}
				style={style}
			>
				<input
					onChange={_.noop}
					{...omitProps(passThroughs, RadioButton, ['children', 'callbackId'])}
					checked={isSelected}
					className={cx('&-native')}
					disabled={isDisabled}
					ref="nativeElement"
					type="radio"
				/>
				<span
					onClick={this.handleSpanClick}
					className={cx('&-visualization-glow')}
				/>
				<span
					onClick={this.handleSpanClick}
					className={cx('&-visualization-container')}
				/>
				<span
					onClick={this.handleSpanClick}
					className={cx('&-visualization-dot')}
				/>
			</span>
		);
	},

	handleClicked(event) {
		const { isDisabled, isSelected, onSelect } = this.props;

		if (!isDisabled && !isSelected) {
			onSelect(true, { event, props: this.props });
			this.nativeElement.focus();
		}
	},
});

export default RadioButton;
