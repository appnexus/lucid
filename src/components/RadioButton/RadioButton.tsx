import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';

import { lucidClassNames } from '../../util/style-helpers';
import { StandardProps, Overwrite } from '../../util/component-types';

const cx = lucidClassNames.bind('&-RadioButton');
const { bool, func, object, string } = PropTypes;

export interface IRadioButtonPropsRaw extends StandardProps {
	/** Indicates whether the component should appear and act disabled by having
	 * a "greyed out" palette and ignoring user interactions.
	 *
	 * @default false
	 */
	isDisabled: boolean;

	/** Indicates that the component is in the "selected" state when true and in
	 * the "unselected" state when false.
	 *
	 * @default false
	 */
	isSelected: boolean;

	/** Optional name for the input element */
	name?: string;

	/** Called when the user clicks on the component or when they press the space
	 * key while the component is in focus, and only called when the component
	 * is in the unselected state.
	 */
	onSelect: (
		isSelected: boolean,
		{
			event,
			props,
		}: {
			event: React.MouseEvent<HTMLInputElement>;
			props: IRadioButtonProps;
		}
	) => void;
}

export type IRadioButtonProps = Overwrite<
	React.DetailedHTMLProps<
		React.HTMLAttributes<HTMLSpanElement>,
		HTMLSpanElement
	>,
	IRadioButtonPropsRaw
>;

/** TODO: Remove this constant when the component is converted to a functional component */
const nonPassthroughs = [
	'callbackId',
	'children',
	'className',
	'isDisabled',
	'isSelected',
	'name',
	'onSelect',
	'style',
];

export const defaultProps = {
	isDisabled: false,
	isSelected: false,
	onSelect: _.noop,
};

export const RadioButton = (props: IRadioButtonProps): React.ReactElement => {
	const {
		className,
		isDisabled,
		isSelected,
		onSelect,
		style,
		...passThroughs
	} = props;

	const nativeElement = React.createRef<HTMLInputElement>();

	function handleClicked(event: React.MouseEvent<HTMLInputElement>): void {
		if (!isDisabled && !isSelected) {
			onSelect(true, { event, props });
			if (nativeElement.current) {
				nativeElement.current.focus();
			}
		}
	}

	function handleSpanClick(e: React.MouseEvent<HTMLInputElement>): void {
		e.preventDefault();
	}

	return (
		<span
			className={cx(
				'&',
				{
					'&-is-disabled': isDisabled,
					'&-is-selected': isSelected,
				},
				className
			)}
			onClick={handleClicked}
			style={style}
		>
			<input
				onChange={_.noop}
				{..._.omit(passThroughs, nonPassthroughs)}
				checked={isSelected}
				className={cx('&-native')}
				disabled={isDisabled}
				ref={nativeElement}
				type='radio'
			/>
			<span onClick={handleSpanClick} className={cx('&-visualization-glow')} />
			<span
				onClick={handleSpanClick}
				className={cx('&-visualization-container')}
			/>
			<span onClick={handleSpanClick} className={cx('&-visualization-dot')} />
		</span>
	);
};

RadioButton.defaultProps = defaultProps;

RadioButton.displayName = 'RadioButton';

RadioButton.peek = {
	description: `\`RadioButton\` is a round two-state toggle used to create \`RadioButtonLabeled\`. It uses a hidden native checkbox control under the hood but leverages other \`HTML\` elements to visualize its state.`,
	notes: {
		overview: `RadioButton is a round two-state toggle. Use \`RadioButtonLabeled\` or \`RadioGroup\` in your applications.`,
		intendedUse: `Used to create \`RadioButtonLabeled\` and \`RadioGroup\`.`,
		technicalRecommendations: `
			- Use the Selected state when a filter or setting will be applied.
			- Use the Unselected state when a filter or setting will not be applied.
			- Any props that are not explicitly defined in \`propTypes\` are passed to the native radio button control.
		`,
	},
	categories: ['controls', 'toggles'],
};

RadioButton.propTypes = {
	/**
		Appended to the component-specific class names set on the root element.
	*/
	className: string,

	/**
		Indicates whether the component should appear and act disabled by having
		a "greyed out" palette and ignoring user interactions.
	*/
	isDisabled: bool,

	/**
		Indicates that the component is in the "selected" state when true and in
		the "unselected" state when false.
	*/
	isSelected: bool,

	/**
	Optional name for the input element.
	*/
	name: string,

	/**
		Called when the user clicks on the component or when they press the space
		key while the component is in focus, and only called when the component
		is in the unselected state.  Signature: \`(true, { event, props }) => {}\`
	*/
	onSelect: func,

	/**
		Passed through to the root element.
	*/
	style: object,
};

export default RadioButton;
