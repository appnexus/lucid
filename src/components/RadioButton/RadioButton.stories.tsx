import _ from 'lodash';
import React from 'react';
import createClass from 'create-react-class';
import RadioButton from './RadioButton';

export default {
	title: 'Controls/RadioButton',
	component: RadioButton,
	parameters: {
		docs: {
			description: {
				component: (RadioButton as any).peek.description,
			},
		},
	},
};

/* Interactive */
export const Interactive = () => {
	const style = {
		listStyleType: 'none',
		display: 'flex',
		alignItems: 'center',
	};

	const Component = createClass({
		getInitialState() {
			return {
				isSelected: false,
			};
		},

		handleSelected(isSelected: boolean) {
			this.setState(
				_.assign({}, this.state, {
					isSelected,
				})
			);
		},

		render() {
			return (
				<ul>
					<li style={style}>
						<label>Enabled</label>
						<RadioButton
							isSelected={this.state.isSelected}
							name='interactive-radio-buttons'
							onSelect={this.handleSelected}
							tabIndex={20}
						/>
					</li>
				</ul>
			);
		},
	});

	return <Component />;
};
Interactive.storyName = 'Interactive';

/* States */
export const States = () => {
	const style = {
		display: 'flex',
		alignItems: 'center',
	};

	const Component = createClass({
		render() {
			return (
				<ul>
					<li style={style}>
						<label>Unselected</label>
						<RadioButton tabIndex={20} />
					</li>
					<li style={style}>
						<label>Selected</label>
						<RadioButton isSelected={true} tabIndex={21} />
					</li>
					<li style={style}>
						<label>Disabled</label>
						<RadioButton isDisabled={true} tabIndex={22} />
					</li>
					<li style={style}>
						<label>Disabled & selected</label>
						<RadioButton isDisabled={true} isSelected={true} tabIndex={23} />
					</li>
				</ul>
			);
		},
	});

	return <Component />;
};
States.storyName = 'States';
