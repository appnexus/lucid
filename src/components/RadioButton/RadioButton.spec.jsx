import _ from 'lodash';
import assert from 'assert';
import React from 'react';
import SyntheticEvent from 'react/lib/SyntheticEvent';
import sinon from 'sinon';
import { mount, shallow } from 'enzyme';

import { common, controls } from '../../util/generic-tests';

import RadioButton from './RadioButton';

describe('RadioButton', () => {
	common(RadioButton);
	controls(RadioButton, {
		callbackName: 'onSelect',
		controlSelector: '.lucid-RadioButton-native',
		eventType: 'click',
	});

	const booleanValues = [true, false];

	describe('props', () => {
		describe('isDisabled', () => {
			it('sets the `disabled` attribute of the native radio button element.', () => {
				_.forEach(booleanValues, (testValue) => {
					const wrapper = shallow(<RadioButton isDisabled={testValue} />);

					assert.equal(wrapper.find('input[type="radio"]').prop('disabled'), testValue);
				});
			});

			it('defaults to `false`.', () => {
				const wrapper = mount(<RadioButton />);

				assert.equal(wrapper.prop('isDisabled'), false);
			});
		});

		describe('isSelected', () => {
			it('sets the `checked` attribute of the native radio button element.', () => {
				_.forEach(booleanValues, (testValue) => {
					const wrapper = shallow(<RadioButton isSelected={testValue} />);

					assert.equal(wrapper.find('input[type="radio"]').prop('checked'), testValue);
				});
			});

			it('defaults to `false`.', () => {
				const wrapper = mount(<RadioButton />);

				assert.equal(wrapper.prop('isSelected'), false);
			});
		});

		describe('onSelect', () => {
			it('defaults to the Lodash `noop` method.', () => {
				const wrapper = mount(<RadioButton />);

				assert.equal(wrapper.prop('onSelect'), _.noop);
			});
		});

		describe('pass throughs', () => {
			it('passes through all props not defined in `propTypes` to the native input.', () => {
				const wrapper = mount(
					<RadioButton className='wut' isDisabled={true} isSelected={true} style={{ fontWeight: 'bold' }} onSelect={_.noop}
							foo={1} bar={2} baz={3} qux={4} quux={5} />
				);
				const nativeProps = wrapper.find('input[type="radio"]').props();

				// It should pass `foo`, `bar`, `baz`, `qux`, and `quux` through
				// to the native input.
				_.forEach(['foo', 'bar', 'baz', 'qux', 'quux'], (prop) => {
					assert(_.has(nativeProps, prop));
				});
			});
		});
	});
});

describe('RadioButton', () => {
	function simulateEvent(reactElement, selector, event) {
		mount(reactElement).find(selector).simulate(event);
	}

	function verifyArguments(event) {
		_.forEach(['', '-native', '-visualization-container', '-visualization-glow', '-visualization-dot'], (classSubString) => {
			const onSelect = sinon.spy();

			simulateEvent(<RadioButton isSelected={false} onSelect={onSelect} />, `.lucid-RadioButton${classSubString}`, event);
			assert.equal(onSelect.args[0][0], true);
			assert(_.last(onSelect.args[0]).event instanceof SyntheticEvent);
		});
	}

	function verifyNoOnSelect(event) {
		_.forEach(['', '-native', '-visualization-container', '-visualization-glow', '-visualization-dot'], (classSubString) => {
			const onSelect = sinon.spy();

			simulateEvent(<RadioButton isSelected={true} onSelect={onSelect} />, `.lucid-RadioButton${classSubString}`, event);
			assert.equal(onSelect.calledOnce, false);
		});
	}

	function verifyOnSelect(event) {
		_.forEach(['', '-native', '-visualization-container', '-visualization-glow', '-visualization-dot'], (classSubString) => {
			const onSelect = sinon.spy();

			simulateEvent(<RadioButton isSelected={false} onSelect={onSelect} />, `.lucid-RadioButton${classSubString}`, event);
			assert(onSelect.calledOnce);
		});
	}

	describe('user clicks on the rendered control', () => {
		it('calls the function passed in as the `onSelect` prop if `isSelected` is false...', () => {
			verifyOnSelect('click');
		});

		it('...and passes along true as the first argument and a React synthetic event as the second argument.', () => {
			verifyArguments('click');
		});

		it('does not call the function passed in as the `onSelect` prop if `isSelected` is true.', () => {
			verifyNoOnSelect('click');
		});
	});

	describe('user taps on the rendered control', () => {
		it('calls the function passed in as the `onSelect` prop if `isSelected` is false...', () => {
			verifyOnSelect('touchend');
		});

		it('...and passes along true as the first argument and a React synthetic event as the second argument.', () => {
			verifyArguments('touchend');
		});

		it('does not call the function passed in as the `onSelect` prop if `isSelected` is true.', () => {
			verifyNoOnSelect('touchend');
		});
	});
});
