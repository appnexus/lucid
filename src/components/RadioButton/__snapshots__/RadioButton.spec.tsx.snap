// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton [common] example testing should match snapshot(s) for 0.interactive 1`] = `
<span
  className="lucid-RadioButton"
  onClick={[Function]}
>
  <input
    checked={false}
    className="lucid-RadioButton-native"
    disabled={false}
    onChange={[Function]}
    tabIndex={20}
    type="radio"
  />
  <span
    className="lucid-RadioButton-visualization-glow"
    onClick={[Function]}
  />
  <span
    className="lucid-RadioButton-visualization-container"
    onClick={[Function]}
  />
  <span
    className="lucid-RadioButton-visualization-dot"
    onClick={[Function]}
  />
</span>
`;

exports[`RadioButton [common] example testing should match snapshot(s) for 1.states 1`] = `
<span
  className="lucid-RadioButton"
  onClick={[Function]}
>
  <input
    checked={false}
    className="lucid-RadioButton-native"
    disabled={false}
    onChange={[Function]}
    tabIndex={20}
    type="radio"
  />
  <span
    className="lucid-RadioButton-visualization-glow"
    onClick={[Function]}
  />
  <span
    className="lucid-RadioButton-visualization-container"
    onClick={[Function]}
  />
  <span
    className="lucid-RadioButton-visualization-dot"
    onClick={[Function]}
  />
</span>
`;

exports[`RadioButton [common] example testing should match snapshot(s) for 1.states 2`] = `
<span
  className="lucid-RadioButton lucid-RadioButton-is-selected"
  onClick={[Function]}
>
  <input
    checked={true}
    className="lucid-RadioButton-native"
    disabled={false}
    onChange={[Function]}
    tabIndex={21}
    type="radio"
  />
  <span
    className="lucid-RadioButton-visualization-glow"
    onClick={[Function]}
  />
  <span
    className="lucid-RadioButton-visualization-container"
    onClick={[Function]}
  />
  <span
    className="lucid-RadioButton-visualization-dot"
    onClick={[Function]}
  />
</span>
`;

exports[`RadioButton [common] example testing should match snapshot(s) for 1.states 3`] = `
<span
  className="lucid-RadioButton lucid-RadioButton-is-disabled"
  onClick={[Function]}
>
  <input
    checked={false}
    className="lucid-RadioButton-native"
    disabled={true}
    onChange={[Function]}
    tabIndex={22}
    type="radio"
  />
  <span
    className="lucid-RadioButton-visualization-glow"
    onClick={[Function]}
  />
  <span
    className="lucid-RadioButton-visualization-container"
    onClick={[Function]}
  />
  <span
    className="lucid-RadioButton-visualization-dot"
    onClick={[Function]}
  />
</span>
`;

exports[`RadioButton [common] example testing should match snapshot(s) for 1.states 4`] = `
<span
  className="lucid-RadioButton lucid-RadioButton-is-disabled lucid-RadioButton-is-selected"
  onClick={[Function]}
>
  <input
    checked={true}
    className="lucid-RadioButton-native"
    disabled={true}
    onChange={[Function]}
    tabIndex={23}
    type="radio"
  />
  <span
    className="lucid-RadioButton-visualization-glow"
    onClick={[Function]}
  />
  <span
    className="lucid-RadioButton-visualization-container"
    onClick={[Function]}
  />
  <span
    className="lucid-RadioButton-visualization-dot"
    onClick={[Function]}
  />
</span>
`;
